OperatorModel - cleanup and testing

DanPuzzuoli opened the pull request:
### Summary

This PR is for finalizing the design of the `OperatorModel` class and adding tests

### Update

This PR is ready for review. Contains:

- Main piece: `OperatorModel` class in `models.py`. 
     - As explained in the documentation, represents a list of operators with signal coefficients, with functionality for evaluating the model, entering a frame, and specifying a frequency cutoff (for things like the RWA). 
     - This file also contains a helper class called `FrameFreqHelper` whose only purpose is to compartmentalize the more technical aspects of the calculations to do with frames/frequencies in `OperatorModel`.
- Secondary piece: Addition of `VectorSignal` to `signals.py`. 
    - Internally, the `OperatorModel` represents the signals as a `VectorSignal` - a class that is the same as `Signal` but has a vector valued envelope and a list of carrier frequencies. 
    - The point of this is to fully abstract the calculation of signal values out of `OperatorModel`, to facilitate future applications in which it is computationally advantageous to compute the values of multiple signals at once (as opposed to iterating through in a list).
    - A `VectorSignal` can be instantiated from a list of `Signal` objects, and the `OperatorModel` will attempt to do this if the signals are passed as a list (preserving the current standard use-case of specifying signals as a list).
- Testing:
  - New test file `pulse_new/models/test_operator_models.py`, containing various tests for `OperatorModel`
- Demo
  - New demo notebook `OperatorModel_demo.ipynb` demonstrating constructing an `OperatorModel`, calling some functions, the effects of entering a frame, and the effects of specifying a frequency cutoff.

~~I think some minor further additions: add additional tests for handling of non-diagonal frame-bases (i.e. evaluation of a model in the basis which the frame is diagonal), as well as the related functionality in `FrameFreqHelper` for specifying which basis to express a "state" in when transforming into/out of the frame. Could add this stuff to the demo too.~~

In addition to the above:
- A bunch of documentation as per @eggerdj 's comments
- Added a check that a specified signal has the same number of entries as the operators
- Removed the function `vector_apply_diag_frame` and moved the code into the only function that was calling it
- Added a section to the `OperatorModel` demo for specifying a `signal_mapping`
- Added a section to `OperatorModel` demo showing helper functions for transforming states into/out of the given frame (which will be used in DE solvers)

Added tests:
- checking warning if length of signal is invalid
- test for setting a `signal_mapping`
- tests for state frame transformations

DanPuzzuoli commented:
Current status: renamed file `models.py` to `operator_models.py` and the `BaseModel` class to `OperatorModel` as the original names seemed too general.

DanPuzzuoli commented:
Adding a bunch of tests for the `FrameFreqHelper`. Tests that remain:
- `evaluate` with frequency cutoffs
- frame transformations for states
- both of the above with various setting of flags for whether the input/output of the function is in the basis in which the frame is diagonal

DanPuzzuoli commented:
This is ready for review (see updated initial comment for details).

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+
+        Args:
+            operators (list): list of Operator objects.
+            signals (list): Specifiable as either a VectorSignal, a list of
+                            Signal objects, or as the inputs to signal_mapping.
```
Perhaps we can add some type hints in the signature of the functions to represent this.
E.g. `Union[List, VectorSignal]`?

DanPuzzuoli replied:
Sounds good - going to add these type hints everywhere to be consistent.

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+            cutoff_freq (float): Frequency cutoff when evaluating the model.
+        """
+
+        self._operators = operators
```
Should we check that operators and signals have the same length?

DanPuzzuoli replied:
Hmm not sure... I think how to handle length checking depends on what type of functionality we want to have for adding terms to a model. That being said we don't have any of that functionality, so perhaps for now I can just add a check when the signals are set that they're the same length as the operators (with the operators being fixed at instantiation).

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+            operators (list): list of Operator objects.
+            signals (list): Specifiable as either a VectorSignal, a list of
+                            Signal objects, or as the inputs to signal_mapping.
+            signal_mapping (Callable): a function returning wither a
+                                       VectorSignal or a list of Signal objects.
```
This docstring could greatly benefit from an example. Looking at the signals setter below it is not quite clear how this mapping should work.

DanPuzzuoli replied:
I've added an example into the demo as well as in the class doc string for how this is meant to behave.

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+
+    @frame_operator.setter
+    def frame_operator(self, frame_operator):
+        """Set the frame operator; must be anti-Hermitian.
```
Can we spend a few more lines on what the frame operator is and what it does?

DanPuzzuoli replied:
I added a sentence pointing to the class doc string for this, where the effects of setting a frame operator are generally explained.

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+              diagonal matrix, and the array gives the diagonal elements.
+        """
+
+        if isinstance(frame_operator, np.ndarray) and frame_operator.ndim == 1:
+            # if 1d array check purely imaginary
+            if np.linalg.norm(frame_operator + frame_operator.conj()) > 10**-10:
+                raise Exception('frame_op must correspond to an anti-Hermitian matrix.')
```
Does `Operator` not work on 1D arrays?

DanPuzzuoli replied:
Currently it doesn't

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+        return np.dot(y, self.evaluate(time, in_frame_basis))
+
+    @property
+    def drift(self):
```
```suggestion
    def drift(self) -> np.array:
```

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+                                 If None, no cutoff is performed.
+        """
+
+        # initial setup of frame operator
```
redundent?
```suggestion
```

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+        im_angular_freqs = 1j * 2 * np.pi * self.carrier_freqs
+        self._S = np.array([w + D_diff for w in im_angular_freqs])
+
+        self._M_cutoff = None
+        if cutoff_freq is not None:
+            self._M_cutoff = ((np.abs(self._S.imag) / (2 * np.pi)) <
+                                            self.cutoff_freq).astype(int)
```
Can we add a comment what this is for?

eggerdj commented on qiskit/providers/aer/pulse_new/models/operator_models.py:
```diff
@@ -0,0 +1,409 @@
+
+
+def vector_apply_diag_frame(t, mats_in_frame_basis, coeffs, S, M_cutoff=None):
+    """Given a list of matrices specified in the frame_basis for a
+    """
```
Incomplete docstring

DanPuzzuoli replied:
I've eliminated this function and just moved the computations into the `FrameFreqHelper.evaluate` method - it was too weird to explain out of context.

eggerdj commented on qiskit/providers/aer/pulse_new/models/signals.py:
```diff
@@ -211,3 +212,96 @@ def signal_add(sig1, sig2):
+        Args:
+            envelope (Callable): function of a single float returning an array.
+            carrier_freqs (array): list of carrier frequences
+            drift_array (array): the drift array
```
Needs more information.

DanPuzzuoli replied:
Tried adding some here - there is also more info in the class doc string

eggerdj reviewed (changes requested):
I think this is looking really good. A few comments:
* In some places a bit more detail in the docstrings would help.
* Some places are missing type hints.

DanPuzzuoli commented:
Summary of changes:
- A bunch of documentation as per @eggerdj 's comments
- Added a check that a specified signal has the same number of entries as the operators
- Removed the function `vector_apply_diag_frame` and moved the code into the only function that was calling it
- Added a section to the `OperatorModel` demo for specifying a `signal_mapping`

Added tests:
- checking warning if length of signal is invalid
- test for setting a `signal_mapping`

eggerdj commented on qiskit/providers/aer/pulse_new/models/signals.py:
```diff
@@ -374,3 +374,106 @@ def signal_add(sig1: Union[BaseSignal, float, int, complex],
+            self.drift_array = drift_array
+
+    @classmethod
+    def from_signal_list(cls, signal_list: List[Signal]):
```
Can we initialize from `BaseSignal` instead? `BaseSignal` defines an `envelop_value` function so this should be fine.
```suggestion
    def from_signal_list(cls, signal_list: List[BaseSignal]):
```

eggerdj reviewed (approved):
Looks good, one minor comment on `VectorSignal`.
